engine/executor/agg_func_prom.go

```diff
@@ -181,3 +183,95 @@ func almostEqual(a, b, epsilon float64) bool {
 	}
 	return diff/math.Min(absSum, math.MaxFloat64) < epsilon
 }
+
+type CallFn func([]int64, []int64, []float64, []float64, int64, *influxql.PromSubCall) (float64, bool)
+
+var promSubqueryFunc map[string]CallFn = make(map[string]CallFn)
+
+func init() {
+	promSubqueryFunc["rate_prom"] = rate
+}
+
+func CalcReduceResult(prevT, currT []int64, prevV, currV []float64, isCounter bool) (int64, int64, float64, float64, float64) {
+	var firstTime, lastTime int64
+	var firstValue, lastValue float64
+	if len(prevT) > 0 {
+		firstTime = prevT[0]
+		firstValue = prevV[0]
+		if len(currT) > 0 {
+			lastTime = currT[len(currT)-1]
+			lastValue = currV[len(currV)-1]
+		} else {
+			lastTime = prevT[len(prevT)-1]
+			lastValue = prevV[len(prevV)-1]
+		}
+	} else {
+		firstTime, lastTime = currT[0], currT[len(currT)-1]
+		firstValue, lastValue = currV[0], currV[len(currV)-1]
+	}
+	reduceResult := lastValue - firstValue
+	if isCounter {
+		prev := firstValue
+		for _, cur := range prevV {
+			if cur < prev {
+				reduceResult += prev
+			}
+			prev = cur
+		}
+		for _, cur := range currV {
+			if cur < prev {
+				reduceResult += prev
+			}
+			prev = cur
+		}
+	}
+	return firstTime, lastTime, firstValue, lastValue, reduceResult
+}
+
+func rate(preTimes []int64, currTimes []int64, preValues []float64, currValues []float64, ts int64, call *influxql.PromSubCall) (float64, bool) {
+	pointCount := len(preTimes) + len(currTimes)
+	if pointCount <= 1 {
+		return 0, false
+	}
+	firstTime, lastTime, firstValue, _, reduceResult := CalcReduceResult(preTimes, currTimes, preValues, currValues, true)
+	if lastTime == firstTime || call.Range.Nanoseconds() == 0 {
+		return 0, false
+	}
+	rangeStart, rangeEnd := ts-call.Range.Nanoseconds(), ts
+
+	// Duration between first/last samples and boundary of range.
+	durationToStart := float64(firstTime-rangeStart) / 1e9
```

Comment:
Use constants to represent repeatedly used values `1e9`

Reply:
This function is derived from the prom source code.